pkgs/cronet_http/lib/src/cronet_client.dart

```diff
@@ -186,26 +212,114 @@ class CronetClient extends BaseClient {
           'HTTP request failed. Client is already closed.', request.url);
     }
 
-    if (_engine == null) {
-      // Create the future here rather than in the [fromCronetEngineFuture]
-      // factory so that [close] does not have to await the future just to
-      // close it in the case where [send] is never called.
-      //
-      // Assign to _engineFuture instead of just `await`ing the result of
-      // `CronetEngine.build()` to prevent concurrent executions of `send`
-      // from creating multiple [CronetEngine]s.
-      _engineFuture ??= CronetEngine.build();
-      try {
-        _engine = await _engineFuture;
-      } catch (e) {
-        throw ClientException(
-            'Exception building CronetEngine: ${e.toString()}', request.url);
-      }
-    }
+    _engine ??= CronetEngine.build();
 
     final stream = request.finalize();
-
     final body = await stream.toBytes();
+    final responseCompleter = Completer<StreamedResponse>();
+    final engine = _engine!._engine;
+
+    late jb.UrlRequest cronetRequest;
+    var numRedirects = 0;
+    StreamController<List<int>>? responseStream;
+    jb.ByteBuffer? byteBuffer;
+
+    // The order of callbacks generated by Cronet is documented here:
+    // https://developer.android.com/guide/topics/connectivity/cronet/lifecycle
+
+    final cronetCallbacks =
+        jb.UrlRequestCallbackProxy_UrlRequestCallbackInterface.implement(
+            jb.$UrlRequestCallbackProxy_UrlRequestCallbackInterfaceImpl(
+      onResponseStarted: (urlRequest, responseInfo) {
+        responseStream = StreamController();
+        final responseHeaders =
+            _cronetToClientHeaders(responseInfo.getAllHeaders());
+        int? contentLength;
+
+        switch (responseHeaders['content-length']) {
+          case final contentLengthHeader?
+              when !_digitRegex.hasMatch(contentLengthHeader):
```

Comment:
[out of scope for this PR] Would it make sense to use `int.tryParse` and report an invalid header if we get back `null`? Can we skip the regex check in this code?

Reply:
Negative numbers are also not allowed so it ends up being two checks. I'm not sure if that is better.

Reply:
Given that I missed that behavior, I have a slight preference for making the negative check explicit with `if (contentLengthHeader.isNegative)` over implicitly checked by the regex, but I don't feel strongly about it.

Reply:
I used the regex because it is a direct transliteration from the [production in the RFC](https://datatracker.ietf.org/doc/html/rfc2616#section-14.13)